### PR TITLE
CORE-20185 Add deduplication ID to the Token Claim API to be used for idempotency

### PR DESCRIPTION
**Pull request by LWogan**
To be supplied by the flow. Required to maintain idempotency in non deterministic flows which may have non deterministic suspension counts when rerunning flows. By forcing the client to provide the deduplication id, the flow engine can replay flopw fibers indefinitely and the token selection worker will still be able to deduplicate requests.

Runtime:
https://github.com/corda/corda-runtime-os/pull/5921

Branched from similar Persist API changes:
https://github.com/corda/corda-api/pull/1574

### CODE REVIEW

**Comment by filipesoliveira** [comment]
The `TokenSelection` currently uses the `externalEventRequestId` to provide idempotency. Don't forget to replace the `externalEventRequestId` if a `deduplication ID` is going to be used instead. 

Link to the affected code: https://github.com/corda/corda-runtime-os/blob/release/os/5.3/components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/handlers/TokenClaimQueryEventHandler.kt#L32

File: ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/token/selection/TokenSelection.java [hunk 1]
```diff
@@ -1,10 +1,13 @@
 package net.corda.v5.ledger.utxo.token.selection;
 
+import net.corda.v5.application.flows.CordaInject;
 import net.corda.v5.base.annotations.DoNotImplement;
 import net.corda.v5.base.annotations.Suspendable;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
+import java.math.BigDecimal;
```
Comment (filipesoliveira) [inline ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/token/selection/TokenSelection.java]: Is this import necessary?

File: ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/token/selection/TokenSelection.java [hunk 2]
```diff
@@ -1,10 +1,13 @@
 package net.corda.v5.ledger.utxo.token.selection;
 
+import net.corda.v5.application.flows.CordaInject;
```
Comment (filipesoliveira) [inline ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/token/selection/TokenSelection.java]: Is this import necessary?

**Review by rick-r3 — DISMISSED** [review]

**Review by filipesoliveira — DISMISSED** [review]

File: ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/token/selection/TokenSelection.java [hunk 3]
```diff
@@ -21,7 +21,14 @@ public interface TokenSelection {
      * If available, a set of tokens will be selected that sum to at least the target amount specified.
      * The tokens will be locked in the cache to prevent other flows from selecting them.
      *
+     * A unique deterministic identifier for a token claim is provided per API call.
+     * Corda may rerun a token claim request on behalf of the calling flow in the event system instability.
+     * This ID will be used to deduplicate TokenClaims in this scenario.
+     * This ID needs to be unique per TokenClaim within a single flow.
+     *
      * @param criteria The {@link TokenClaimCriteria} used to select tokens.
```
Comment (sprintengine) [inline ledger/ledger-utxo/src/main/java/net/corda/v5/ledger/utxo/token/selection/TokenSelection.java:29]: I'd wonder about adding an example using a static constant or similar to Ben's https://github.com/corda/corda-api/pull/1574/files#r1538964821 but I wouldn't let it hold up the PR

**Review by rick-r3 — APPROVED** [review]

**Pull request merged**